xosview.1
'\" t
.\" @(#)xosview.1	1.8.4 1/99 "
.TH XOSVIEW 1.8.4 "$Date: 2008/02/28 23:43:06 $"
.UC
.SH NAME
xosview \- X based system monitor
.SH SYNOPSIS
xosview [options]
.\"  ================  Macros  =============================
.\"  First, let's define some handy roff macros.
.\"  A macro begins with .de <xx> where one will invoke this macro with .xx
.\"  The macro definition ends with the .. line.
.\"  I don't know what macro abbreviations are free, so I just chose a few,
.\"  and haven't noticed a problem so far!  bgrayson

.\"  There are several paragraphs that are repeated in the resource section.
.\"  Rather than typing the whole stuff out each time, we define a few macros.

.\"  The .pp macro takes a single argument (net, disk, etc), and
.\" prints a paragraph description of a Priority resource.  Only the
.\" header (xosview*diskPriority: \fIpriority\fP) needs to be specified in
.\" addition to the .pp macro.
.\"  Usage:   .pp load
.de pp
xosview*\\$1Priority: \fIpriority\fP
.RS
This number (which must be an integer >= 1) sets the number of tenths of
a second that the \\$1 meter waits between updates.  A value of 1 has xosview
update the meter 10 times per second (the fastest).  A value of 600 would
cause xosview to update the meter once a minute.
.RE
..
.\"  The .dc macro is similar to the .pp macro, except that it is for
.\" the decay resource paragraphs.
.\"  Usage:  .dc net
.de dc
xosview*\\$1Decay: (True or False)
.RS
If True then the \\$1 meter will be split vertically in two.  The top
half will show the instantaneous state, while the bottom half will
display a decaying average of the state.
.RE
..
.\" The .dg macro is similar to the .dc macro, except that it is for
.\" the scrolling graph resource paragraphs.
.\" Usage:  .dg net
.de dg
xosview*\\$1Graph: (True or False)
.RS
If this is set to True then the \\$1 meter will be drawn as a horizontally
scrolling bargraph showing the state value verses time.
.RE
..
.\"  The .pm macro is for ``plus/minus'' -- for the
.\"  enable/disable command-line options.
.de pm
-\\$1
.RS
This option overrides the xosview*\\$1 resource.  It is
equivalent to setting xosview*\\$1 to "False".
.RE
+\\$1
.RS
This option overrides the xosview*\\$1 resource.  It is
equivalent to setting xosview*\\$1 to "True".
.RE
..
.\"  The .xt macro is for ``Xresource true'' -- for enabling a
.\"  meter, like the .pm macro/-+foo.
.de xt
xosview*\\$1: (True or False)
.RS
If True then xosview will display a \\$1 meter.
.RE
..
.\"  The .uf macro is for the ``used format'' stuff.
.de uf
xosview*\\$1UsedFormat:	  (float, percent or autoscale)
.RS
This resource tells xosview how to display "used" labels.  The formats work
as follows:

\fBfloat\fP:
.RS
Display the value as a floating point number.
.RE
\fBpercent\fP:
.RS
Display the value as a percentage of the total.
.RE
\fBautoscale\fP:
.RS
Display the absolute value and automatically print the units (K, M, or G) as
appropriate.
.RE
.RE
..
.\"  Define a color macro for the various xosview*fooXXXColor: resources.
.\"  Usage:  .cc swap Used used "swap space"  <--  keep 'swap
\"                                                       space' as one arg.
.de cc
xosview*\\$1\\$2Color: \fIcolor\fP
.RS
The \\$1 meter will use this color to display the \\$3 field.
.RE
..
.\"  ================  End of Macros  =============================
.SH DESCRIPTION
\fIxosview\fP is a monitor which displays the status of several system based
parameters.  These include CPU usage, load average, memory usage, swap
space usage, network usage, interrupts, and serial port status.  Each of
these is displayed as a horizontal bar which is separated into color coded
regions.  Each region represents a percentage of the resource which is being
put to a particular use.  Typing a 'q' in the window will terminate xosview.

At the moment \fIxosview\fP runs on eight platforms (Linux,
NetBSD, FreeBSD, OpenBSD, some Solaris systems, IRIX 6.5, HPUX and GNU).
Not all of the meters described below are supported on all
platforms.  Some of the meters may appear different depending upon the
platform \fIxosview\fP is running on.  Note that *BSD is used as
an abbreviation for all three of the free BSD operating systems
(NetBSD, FreeBSD, OpenBSD).

\fBLoad\fP
   all platforms  :  load

\fBCPU Usage\fP
   Linux  : user | nice | system | iowat | irq | softirq |
            stolen | guest | nice guest | idle
   *BSD   : user | nice | system | interrupt | idle
   IRIX   : user | system | interrupt | wait | idle
   HPUX   : user | nice | system | interrupt | idle

\fBMemory Usage\fP
   Linux  : used | shared | buffers | free
   NetBSD : active | inactive | wired | free
   FreeBSD: active | inactive | wired | buffer | free
   OpenBSD: active | inactive | wired | free
   IRIX   : kernel | fs   | user    | free
   HPUX   : text | used   | other   | free
   GNU    : active | inactive | wired | free

\fBSwap Usage\fP
   all platforms except IRIX: used | free

\fBDisk Usage (bytes/sec)\fP
   Linux  : in  |  out  | idle
   *BSD   : transferred | idle
   HPUX   : < Not Supported >

\fBPage Swapping from/to disk\fP
   all platforms except IRIX  : in | out | idle

\fBNET Usage\fP
    bytes/sec
    Linux  : in | out | idle
    *BSD   : in | out | idle
    HPUX   : < Not Supported >

\fBGFX Usage\fP
    swapbuffers/sec
    IRIX   : number of swapbuffers

\fBInterrupts\fP
    "leds" which blink when an interrupt occurs.
    Linux  : IRQs 0 - 64 (per cpu if configured)
    *BSD   : IRQs 0 - 15 (or fewer, depending on architecture)
    HPUX   : < Not Supported >

\fBInterrupt rate\fP
    interrupts per second
    Linux  : < Not Supported >
    *BSD   : interrupts | free
    HPUX   : < Not Supported >

\fBSerial Port Status\fP
    "leds" which show the serial port parameters.
    Linux  : Both the Line Status Register
             (LSR) and Modem Status Register
             (MSR) are displayed.
    *BSD   : < Not Supported >
    HPUX   : < Not Supported >

\fBBattery Level\fP
    Shows the current condition of the
    battery (for laptops).
    Linux  : %left | %used
    NetBSD : %left | %used
    FreeBSD: < Not Supported >
    OpenBSD: %left | %used
    HPUX   : < Not Supported >

\fBRAID status\fP
    Shows the state of disks contained in your (software)
    RAID1/4/5 array(s) and the rebuild state of the
    array(s).
    Linux  : Disk0 Disk1 Disk2 .. DiskN | Rebuild Progress

.SH OPTIONS

Most of these command line options are just a convenient way to set one or
more of xosview's resources.  Please see the \fBRESOURCES\fP section for
more details on what the resource is for.

-v
.RS
Displays the version number.
.RE

-name \fIname\fP
.RS
Sets the Resource name xosview will use (same as the -name option
for typical X applications like xterm).  When performing resource
lookups, xosview will look under \fIname\fP, and then under
``xosview''.  For an example, perform the following (as
documented in README.netbsd):
.RS
  xrdb -merge Xdefaults.stipple
  xosview -name xosvstipple &
  xosview -name xosvstipplebw &
.RE

.RE

-display \fIdisplay\fP
.RS
Sets the X display to \fIdisplay\fP.  This option overrides the
xosview*display resource.
.RE

-font \fIfont\fP
.RS
Specifies the font xosview will use for labels.  This option overrides the
xosview*font resource.
.RE

-title \fItitle\fP
.RS
This option sets the name xosview will tell the window manager to use for
the X window and icon.  This option overrides the xosview*title resource.
.RE

-geometry \fIgeometry_string\fP
.RS
Sets the X geometry to \fIgeometry_string\fP.  This option overrides the
xosview*geometry resource.
.RE

.\"  Handle -/+captions
.pm captions

.\"  Handle -/+labels
.pm labels

-usedlabels
.RS
This option overrides the xosview*usedlabels resource.  It is equivalent to
setting xosview*usedlabels to "False".
.RE

+usedlabels
.RS
This option overrides the xosview*usedlabels resource.  It is equivalent to
setting xosview*usedlabels to "True".
.RE

.\"  Handle -/+cpu
.pm cpu

.\"  Handle -/+load
.pm load

.\"  Handle -/+mem
.pm mem

.\"  Handle -/+swap
.pm swap

.\"  Handle -/+battery
.pm battery

.\"  Handle -/+gfx
.pm gfx

.\"  Handle -/+net
.pm net

-network \fImaxbandwidth\fP
.RE
-networkBW \fImaxbandwidth\fP
.RE
-networkBandwidth \fImaxbandwidth\fP
.RS
These options override the xosview*netBandwidth resource.  They cause
xosview to display a meter that will shows network usage, with a maximum
bandwidth of \fBmaxbandwidth\fP.  Notice that setting the bandwidth to
0 no longer disables the meter -- use the ``-net'' option instead.
.RE

.\"  Handle -/+page
.pm page

-pagespeed \fIval\fP
.RS
This option overrides the xosview*pageBandWidth resource.  The resource
xosview*pageBandWidth will be set to \fIval\fP.
.RE

.\"  Handle -/+disk
.pm disk

.\"  Handle -/+ int
.pm int

-ints +ints
.RE
-interrupts +interrupts
.RS
Equivalent to -int and +int.
.RE

.pm irqrate

-intrate +intrate
.RS
Equivalent to -irqrate and +irqrate.
.RE

.\"  Handle -/+lmstemp
.pm lmstemp
.RE

-xrm \fIresource_string\fP
.RS
This switch allows any of xosview's resources to be set on the command line.
An example of how the xosview*memFreeColor could be set using this option is
shown below (Note the use of " to prevent the shell from expanding
\'*\' or from creating two separate arguments, \'xosview*memfreeColor:\'
and \'purple\'):
.RS
-xrm "xosview*memFreeColor: purple"
.RE
.RE

.SH X RESOURCES

The following is a list of X resources supported by \fIxosview\fP.  Each has
a default value assigned to it.  These values can be found in the file
Xdefaults which can be obtained in the source distribution of xosview.
They can be overridden in the usual places (/usr/lib/X11/app-defaults/XOsview,
$HOME/.Xdefaults, etc.).

It should be noted that it is OK to have a resource defined for a port of
xosview that does not support the feature the resource configures.  Xosview
will simply ignore the resources that are set for it but not supported on
a given platform.


\fBGeneral Resources\fP


xosview*title: \fIname\fP
.RS
The string that xosview will use for the X window title.  Normally xosview
will use 'xosview@machine_name' for a title.  This resource overrides the
default behavior.
.RE

xosview*geometry: \fIgeometry_string\fP
.RS
This is a standard X geometry string that defines the size and location of
the X window used by xosview.
.RE

xosview*display: \fIname\fP
.RS
The name of the display where xosview will contact the X server for drawing
its window.
.RE

xosview*pixmapName: \fIname\fP
.RS
The filename of an X pixmap (xpm) file for use as a background
image.
.RE

xosview*transparent: (True or False)
.RS
If True set the window background to transparent.  EXPERIMENTAL.
.RE

xosview*captions: (True or False)
.RS
If True then xosview will display meter captions.
.RE

xosview*labels: (True or False)
.RS
If True then xosview will display meter labels.
.RE

xosview*meterLabelColor: \fIcolor\fP
.RS
The color to use for the meter labels.
.RE

xosview*usedlabels: (True or False)
.RS
If True then xosview will display labels that show the percentage of the
resource (or absolute amount, depending on the meter) being used.  This
option requires that the labels option also be set to True.
.RE

xosview*usedLabelColor: \fIcolor\fP
.RS
The color to use for "used" labels.
.RE

xosview*borderwidth:  \fIwidth\fP
.RS
The width of the border for the xosview window.
.RE

xosview*font: \fIfont\fP
.RS
This is the font that xosview will use.
.RE

xosview*background: \fIcolor\fP
.RS
This is the color that will be used for the background.
.RE

xosview*foreground: \fIcolor\fP
.RS
This is the color that will be used for the foreground.
.RE

xosview*enableStipple:	(True or False)
.RS
Change to true to try stipple support.  This is primarily for users
stuck with 1-bit monitors/display cards.  Try setting enableStipple
true.  Please give us feedback on this, if you use it.  It needs
some more work, but no one has given us any feedback so far.
.RE

xosview*graphNumCols: \fInumber\fP
.RS
This defines the number of sample bars drawn when a meter is in scrolling
graph mode. This also has the side-effect of defining the width of the
graph columns. This is only used by meters which have graph mode enabled.
.RE


\fBLoad Meter Resources\fP


.\"  Do the load: True resource.
.xt load

xosview*loadWarnColor: \fIcolor\fP
.RS
This is the color that the load meter will use once the load average is
greater than the loadWarnThreshold.
.RE

xosview*loadCritColor: \fIcolor\fP
.RS
This is the color that the load meter will use once the load average is
greater than the loadCritThreshold (peaked).
.RE

xosview*loadProcColor: \fIcolor\fP
.RS
This is the color that the load meter will use to display the load average
when it is less than the loadWarnThreshold.
.RE

.\"  loadIdleColor
.cc load Idle idle

.\"  Do the priority resource
.pp load

xosview*loadWarnThreshold: \fIint\fP
.RS
This number (which must be an integer >= 1) sets the value at which
the loadmeter changes its status and color from "normal" to "alarm".
The default value is 2.
.RE

xosview*loadCritThreshold: \fIint\fP
.RS
This number (which must be an integer >= 1) sets the value at which
the loadmeter changes its status and color from "normal" to "alarm".
It is also the value that will be used for the peak of the meter.  The
meter will still show the total load in the used level when peaked.
.RE

xosview*loadDecay: (True or False)
.RS
You should probably leave this at the default value (False).  The load
is already a time-averaged value!
.RE

.dg load

.\"  loadUsedFormat resource
.uf load

xosview*loadCpuSpeed:   (True or False)
.RS
Display the current CPU speed in the load meter.
.RE

\fBCPU Meter Resources\fP


xosview*cpu: (True or False)
.RS
If True then xosview will display a cpu meter.  On linux or IRIX 6.5
SMP machines, the resource cpuFormat decides which meters are
created for multiple CPUs.
.RE

.\"  cpuUserColor, cpuNiceColor, cpuSystemColor, cpuWaitColor, cpuInterruptColor, cpuSoftIntColor, cpuStolenColor, cpuGuestColor, cpuNiceGuestColor, cpuFreeColor
.cc cpu User "cpu user time"
.cc cpu Nice "cpu nice time"
.cc cpu System "cpu system time"
.cc cpu Wait "cpu IO wait time"
.cc cpu Interrupt "cpu IRQ time"

CHANGES
  /*  $Id: CHANGES,v 1.57 2006/03/01 05:36:25 romberg Exp $  */

Changes since xosview-1.8.3

  - Modify linux serialmeter startup to allow for a port number instead of
    True/False.  Patch provided by Horst Wente.

  - netIface support for NetBSD.  Patch from  Bernd Ernesti.

Changes since xosview-1.8.2
 - NFSDStats ans NFSStats Graph, Decay and UsedMode are now configurable

 - SunOS updates, compiles now cleanly with gcc 3.3.2

 - Fix for linux intmeter for interrupts > 16 thanks to John Adams

 - Small patch from Gunnar Wolf for lmstemp (change int to double)

 - Fix battery meter for case where lack of APM report -1%.  Thanks
   to Gunnar Wolf.

 - Allow netmeter to specify, via the Xresource xosview*netIface,
   which network interface it will monitor.  Patch thanks to Gunnar Wolf.

 - Add patch for additional cpu fields in linux for 2.6 kernel.  Patch thanks
   to Hendrik Fehr.

 - Add various pathes from the NetBSD project.  Pathes contributed by Roland
   Illig.

 - Fix bug 854215 (-v and --version still needed to connect to X server)

Changes since xosview-1.8.1
 - Applied sourceforge patch 927112 from Russell Reed which fixes
   the swapmeter when USESYSCALLS is defined.

 - Applied sourceforge patch 959433 which adds acpi support to
   the linux battery meter.

 - Applied a patch similar to the sourceforge patch 586184 submitted
   by Mark Guertin.

 - Linux diskmeter now uses vmstat and thus works with 2.6.
   But the units problem (block size) probably still remains.

 - Linux pagemeter should now work with 2.6 kernels

 - Merge in the fedora-2 nfs patch.

 - Merge in the fedora-2 strip patch.

 - Merge in the fedora-2 proc patch.

 - Merge in the fedora-2 s390 patch.

 - Merge in the fedora-2 non-i386 patch.

 - Merge in the fedora-2 linux-2.6 kernel (related) patch.

 - Add an autogen.sh script.  This script runs autoconf to create
   the configure script.

Changes since xosview-1.8.0

- Update linux port to build with gcc 3.2 and 3.3 (should still work
  with 2.95.

- Add -Wno-deprecated for gcc builds.  This avoids warnings about
  the new C++ header files.  This should be fixed for the next release.

- Change MAX_PROC_LENGTH in linux/cpumeter to 4096 to avoid
  an infinite loop with redhat 7.3

Changes since xosview-1.7.3

- Add Mike Butler's patch which fixes an infinite loop in
  the linux netmeter when a pcmcia card is ejected.

- Add Leopold Toetsch's linux lmstemp meter.

- Linux port (and maybe others) now works with gcc-3.0.4

- added IRIX 6.5 support

- Updated NetBSD support, based on patches from Bernd Ernesti, to
  use new CPU meter sysctl interface.

Changes since xosview-1.7.2

- IRQ rate meter for BSDs added.

- FreeBSD 4.* fixes by David O'Brien

Changes since xosview-1.7.1

- Linux now has a RAID meter.  See README.linux for how to set it up.

  BSDI support, courtesy of Tomer Klainer.
  Fixed memstat module for linux-2.2

  Changes from Thomas Waldmann (ThomasWaldmann@gmx.de)
- Fixed cosmetical bug in xosview:
  if you tried to make the xosview window height too small, it looked quite
  strange (because the height value for the meters got negative)

- Fixed cosmetical bug in disk meter:
  at first call, it displayed negative values, overwriting the "DISK" label

- Cosmetical fix:
  The titles were at x=0 - directly at the window border.
  Fixed it to show 1 pixel background between window border and title texts.
  Also moved "used" displays 2 pixels to the right, so that there is also 1
  pixel between the titles and the used display texts.

- Cosmetical fix: Xdefaults and Xdefaults.stipple file now looks nicer.

- Cosmetical fix: load meter sometimes went beyond the right border (I think
  when showing values >>20). Changed load meter scaling to be really dynamic
  (no upper bound for total, but total doesn't become less than 1.0).
  Changed "used" display to show load relative to 1.0 (so it now just shows
  the load value ;-).

- New feature: load meter now has not only a warning color (and threshold),
  but also a (higher) critical color (and threshold).

- Added info to disk meter problem documentation (see source):
  on my system, it showed only the half of the correct values, factor
  for SW-RAID5 md devices seems to be 2048 (md0), while standard ext2fs
  factor is 1024 on my system, too (hda).

  So there is a problem. Solution ???

- Fixed bitmeter to calculate graphics coordinates like fieldmeter

- MemMeter showed strange values when starting vmware (and having memstat
  module loaded): "used" memory got negative !! Looking at the values, I had
  the impression that "shared" memory is also counted as cache memory.
  I hopefully fixed that by subtracting shared from cache - looks better now.

Changes since xosview-1.7.0
  Optimised scrolling graph to use XCopyArea.
  Fixes to linux diskmeter.
  Fixed all (I hope) linux code which parsed /proc/stat with too
    small a buffer.

Changes since xosview-1.6.2
  Added a linux diskmeter which *should* be similar to the BSD version
    but which also differentiates between reading and writing. Linux
    looks like it actually has more stats (per disk input/output) so
    this could be expanded upon.
  Better *BSD support:  OpenBSD intmeter, FreeBSD diskmeter,
    NetBSD battery meter.
  Some Solaris support.

Changes since xosview-1.6.1
  Check for and set -D__SMP__ if the linux memstat module is built on a
    SMP system.
  New option to omit the meter captions (+-captions) added by
    Christian 'Dr. Disk' Hechelmann <drdisk@tc-gruppe.de>
  New graphing meter display mode added.
  Fixes to work with linux/alpha.
  BSD directories merged -- now a single bsd subdir shared by
    NetBSD, FreeBSD, and OpenBSD.

Changes since xosview-1.6.0:
  Linux netmeter for kernels 1.1 and above no longer hangs after
    the packet count gets too high.
  Interrupt meter now supports priority settings under *BSD and Linux.
  Better UVM detection for *BSD.
  Code cleanup/removal of unneeded #include's for *BSD.


xosview.cc
//
//  Copyright (c) 1994, 1995, 2002, 2006, 2015
//  by Mike Romberg ( mike-romberg@comcast.net )
//
//  This file may be distributed under terms of the GPL
//
#include "xosview.h"
#include "meter.h"
#include "x11font.h"
#ifdef HAVE_XFT
#include "xftfont.h"
#endif
#include "MeterMaker.h"
#include "strutil.h"
#if (defined(XOSVIEW_NETBSD) || defined(XOSVIEW_FREEBSD) || defined(XOSVIEW_OPENBSD))
#include "kernel.h"
#endif

#include <algorithm>

#include <unistd.h>



static const char * const versionString = "xosview version: " PACKAGE_VERSION;
static const char NAME[] = "xosview@";


double XOSView::MAX_SAMPLES_PER_SECOND = 10;


XOSView::XOSView(int argc, char *argv[])
    : XWin(), xrm(Xrm("xosview", iname(argc, argv))),
      expose_flag_(false), exposed_once_flag_(false) {

    // Check for version arguments first.  This allows
    // them to work without the need for a connection
    // to the X server
    checkVersion(argc, argv);

    setDisplayName(xrm.getDisplayName(argc, argv));

    openDisplay();  //  So that the Xrm class can contact the display for its
                    //  default values.

    //  The resources need to be initialized before calling XWinInit, because
    //  XWinInit looks at the geometry resource for its geometry.  BCG
    xrm.loadAndMergeResources(argc, argv, display());
    XWinInit (argc, argv);

    setSleepTime();

    loadResources(argc, argv); // General var init

    setEvents();  //  set up the X events

    checkOverallResources(); // see if legends are to be used

    createMeters(); // add in the meters

    // determine the width and height of the window then create it
    // These *HAVE* to come before the resource checking
    // because checking resources alloc colors
    // And there is no display set in the graphics
    // until init() does it's thing.
    figureSize();
    Xrm::opt geom = xrm.getResource("geometry");
    init(argc, argv, getResourceOrUseDefault("pixmapName", ""),
      geom.second, !geom.first);

    checkMeterResources(); //  Have the meters re-check the resources.

    title( winname() );
    iconname( winname() );
    dolegends();
    resize();
}

void XOSView::checkVersion(int argc, char *argv[]) const {
    for (int i = 0 ; i < argc ; i++)
        if ((util::tolower(argv[i]) == "-v")
          || (util::tolower(argv[i]) == "--version")) {
            std::cout << versionString << std::endl;
            exit(0);
        }
}

int XOSView::findx(XOSVFont &font){
    if ( legend_ ){
        if ( !usedlabels_ )
            return font.textWidth( "XXXXXXXXXXXXXXXXXXXXXXXX" );
        else
            return font.textWidth( "XXXXXXXXXXXXXXXXXXXXXXXXXXXXX" );
    }
    return 80;
}

int XOSView::findy(XOSVFont &font){
    if ( legend_ )
        return 10 + font.textHeight() * _meters.size() * ( caption_ ? 2 : 1 );

    return 15 * _meters.size();
}

void XOSView::figureSize(void) {
    std::string fname = getResource("font");
    logDebug << "Font name: " << fname << std::endl;
#ifdef HAVE_XFT
    X11ftFont font(display(), fname);
#else
    X11Font font(display(), fname);
#endif
    if (!font)
        logFatal << "Could not load font: " << fname << std::endl;

    if ( legend_ ){
        if ( !usedlabels_ )
            xoff_ = font.textWidth( "XXXXX" );
        else
            xoff_ = font.textWidth( "XXXXXXXXX" );

        yoff_ = caption_ ? font.textHeight() + font.textHeight() / 4 : 0;
    }
    static bool firsttime = true;
    if (firsttime) {
        firsttime = false;
        width(findx(font));
        height(findy(font));
    }
}

void XOSView::checkMeterResources( void ){
    for (size_t i = 0 ; i < _meters.size() ; i++)
        _meters[i]->checkResources();
}

int XOSView::newypos( void ){
    return 15 + 25 * _meters.size();
}

void XOSView::dolegends( void ){
    for (size_t i = 0 ; i < _meters.size() ; i++) {
        _meters[i]->docaptions(caption_);
        _meters[i]->dolegends(legend_);
        _meters[i]->dousedlegends(usedlabels_);
    }
}

void XOSView::checkOverallResources() {
    //  Check various resource values.

    //  Set 'off' value.  This is not necessarily a default value --
    //    the value in the defaultXResourceString is the default value.
    usedlabels_ = legend_ = caption_ = 0;

    // use captions
    if ( isResourceTrue("captions") )
        caption_ = 1;

    // use labels
    if ( isResourceTrue("labels") )
        legend_ = 1;

    // use "free" labels
    if ( isResourceTrue("usedlabels") )
        usedlabels_ = 1;
}

std::string XOSView::winname( void ){
    char host[100];
    std::string hname("unknown");
    if (gethostname( host, 99 )) {
        logProblem << "gethostname() failed" << std::endl;
    }
    else {
        host[99] = '\0';  // POSIX.1-2001 says truncated names not terminated
        hname = std::string(host);
    }
    std::string name = std::string(NAME) + hname;
    return getResourceOrUseDefault("title", name);
}


void  XOSView::resize( void ){
    int spacing = vspacing_+1;
    int topmargin = vmargin_;
    int rightmargin = hmargin_;
    int newwidth = width() - xoff_ - rightmargin;
    size_t nmeters = _meters.size();
    nmeters = nmeters ? nmeters : 1; // don't divide by zero
    int newheight =
        (height() -
          (topmargin + topmargin + (nmeters-1)*spacing + nmeters*yoff_)
            ) / nmeters;
    newheight = (newheight >= 2) ? newheight : 2;

    for (size_t i = 0 ; i < _meters.size() ; i++) {
        _meters[i]->resize(xoff_,
          topmargin + (i + 1) * yoff_ + i * (newheight+spacing),
          newwidth, newheight);
    }
}


XOSView::~XOSView( void ){
    logDebug << "deleting " << _meters.size() << " meters..." << std::endl;
    for (size_t i = 0 ; i < _meters.size() ; i++)
        delete _meters[i];
    _meters.resize(0);
}

void XOSView::reallydraw( void ){
    logDebug << "Doing draw." << std::endl;
    g().clear();

    for (size_t i = 0 ; i < _meters.size() ; i++)
        _meters[i]->draw(g());

    g().flush();

    expose_flag_ = false;
}

void XOSView::draw ( void ) {
    if (hasBeenExposedAtLeastOnce() && isAtLeastPartiallyVisible())
        reallydraw();
    else {
        if (!hasBeenExposedAtLeastOnce()) {
            logDebug << "Skipping draw:  not yet exposed." << std::endl;
        } else if (!isAtLeastPartiallyVisible()) {
            logDebug << "Skipping draw:  not visible." << std::endl;
        }
    }
}

void XOSView::run( void ){
    int counter = 0;

    while( !done() ){
        checkevent();
        if (done()) {
            logDebug << "checkevent() set done" << std::endl;
            break; // XEvents can set this
        }

        if (_isvisible){
            for (size_t i = 0 ; i < _meters.size() ; i++) {
                if ( _meters[i]->requestevent() )
                    _meters[i]->checkevent();
            }

            g().flush();
        }
#ifdef HAVE_USLEEP
        /*  First, sleep for the proper integral number of seconds --
         *  usleep only deals with times less than 1 sec.  */
        if (sleeptime_)
            sleep((unsigned int)sleeptime_);
        if (usleeptime_)
            usleep( (unsigned int)usleeptime_);
#else
        usleep_via_select ( usleeptime_ );
#endif
        counter = (counter + 1) % 5;
    }
    logDebug << "leaging run()..." << std::endl;
}

void XOSView::usleep_via_select( unsigned long usec ){
    struct timeval time;

    time.tv_sec = (int)(usec / 1000000);
    time.tv_usec = usec - time.tv_sec * 1000000;

    select( 0, 0, 0, 0, &time );
}

void XOSView::keyPressEvent( XKeyEvent &event ){
    char c = 0;
    KeySym key;

    XLookupString( &event, &c, 1, &key, NULL );

    if ( (c == 'q') || (c == 'Q') )
        done(true);
}

void XOSView::checkArgs (int argc, char** argv) const {
    //  The XWin constructor call in the XOSView constructor above
    //  modifies argc and argv, so by this
    //  point, all XResource arguments should be removed.  Since we currently
    //  don't have any other command-line arguments, perform a check here
    //  to make sure we don't get any more.
    if (argc == 1)
        return;  //  No arguments besides X resources.

    //  Skip to the first real argument.
    argc--;
    argv++;
    while (argc > 0 && argv && *argv) {
        switch (argv[0][1]) {
        case 'n': //  Check for -name option that was already parsed
                  //  and acted upon by main().
            if (util::tolower(*argv) == "-name") {
                argv++;	//  Skip arg to -name.
                argc--;
            }
            break;
#if (defined(XOSVIEW_NETBSD) || defined(XOSVIEW_FREEBSD) || defined(XOSVIEW_OPENBSD))
        case 'N':
            if (strlen(argv[0]) > 2)
                SetKernelName(argv[0]+2);
            else {
                SetKernelName(argv[1]);
                argc--;
                argv++;
            }
            break;
#endif
            /*  Fall through to default/error case.  */
        default:
            logEvent << "Ignoring unknown option '" << argv[0] << "'.\n";
            break;
        }
        argc--;
        argv++;
    }
}

void XOSView::exposeEvent( XExposeEvent &event ) {
    _isvisible = true;
    if ( event.count == 0 ) {
        expose_flag_ = true;
        exposed_once_flag_ = true;
        draw();
    }
    logDebug << "Got expose event." << std::endl;
    if (!exposed_once_flag_) {
        exposed_once_flag_ = 1;
        draw();
    }
}

void XOSView::resizeEvent( XEvent &e ) {
    g().resize(e.xconfigure.width, e.xconfigure.height);
    resize();
    expose_flag_ = true;
    exposed_once_flag_ = true;
    draw();
}


void XOSView::visibilityEvent( XVisibilityEvent &event ){
    _ispartiallyvisible = false;
    if (event.state == VisibilityPartiallyObscured){
        _ispartiallyvisible = true;
    }

    if (event.state == VisibilityFullyObscured){
        _isvisible = false;
    }
    else {
        _isvisible = true;
    }
    logDebug << "Got visibility event; " << _ispartiallyvisible
             << " and " << _isvisible << std::endl;
}

void XOSView::unmapEvent( XUnmapEvent & ){
    _isvisible = false;
}

double XOSView::maxSampRate(void) {
    return MAX_SAMPLES_PER_SECOND;
}

std::string XOSView::iname(int, char **argv) {
    /*  Icky.  Need to check for -name option here.  */
    char** argp = argv;
    const char* instanceName = "xosview";	// Default value.
    while (argp && *argp) {
        if (std::string(*argp) == "-name")
            instanceName = argp[1];
        argp++;
    }  //  instanceName will end up pointing to the last such -name option.

    return instanceName;
}

void XOSView::setSleepTime(void) {
    MAX_SAMPLES_PER_SECOND = util::stof(getResource("samplesPerSec"));
    if (!MAX_SAMPLES_PER_SECOND)
        MAX_SAMPLES_PER_SECOND = 10;

    usleeptime_ = (unsigned long) (1000000/MAX_SAMPLES_PER_SECOND);
    if (usleeptime_ >= 1000000) {
        /*  The syscall usleep() only takes times less than 1 sec, so
         *  split into a sleep time and a usleep time if needed.  */
        sleeptime_ = usleeptime_ / 1000000;
        usleeptime_ = usleeptime_ % 1000000;
    } else {
        sleeptime_ = 0;
    }
#if (defined(XOSVIEW_NETBSD) || defined(XOSVIEW_FREEBSD) || defined(XOSVIEW_OPENBSD))
    BSDInit();	/*  Needs to be done before processing of -N option.  */
#endif
}

void XOSView::loadResources(int argc, char **argv) {
    hmargin_  = util::stoi(getResource("horizontalMargin"));
    vmargin_  = util::stoi(getResource("verticalMargin"));
    vspacing_ = util::stoi(getResource("verticalSpacing"));
    hmargin_  = std::max(0, hmargin_);
    vmargin_  = std::max(0, vmargin_);
    vspacing_ = std::max(0, vspacing_);

    checkArgs (argc, argv);  //  Check for any other unhandled args.
    xoff_ = hmargin_;
    yoff_ = 0;
    appName("xosview");
    _isvisible = false;
    _ispartiallyvisible = false;
    expose_flag_ = true;
    exposed_once_flag_ = false;
}

void XOSView::setEvents(void) {
    addEvent( new Event( this, ConfigureNotify,
        (EventCallBack)&XOSView::resizeEvent ) );
    addEvent( new Event( this, Expose,
        (EventCallBack)&XOSView::exposeEvent ) );
    addEvent( new Event( this, KeyPress,
        (EventCallBack)&XOSView::keyPressEvent ) );
    addEvent( new Event( this, VisibilityNotify,
        (EventCallBack)&XOSView::visibilityEvent ) );
    addEvent( new Event( this, UnmapNotify,
        (EventCallBack)&XOSView::unmapEvent ) );
}

void XOSView::createMeters(void) {
    MeterMaker mm(this);
    mm.makeMeters();
    for (int i = 1 ; i <= mm.n() ; i++)
        _meters.push_back(mm[i]);

    if (_meters.size() == 0)
        logProblem << "No meters were enabled." << std::endl;
}

bool XOSView::isResourceTrue( const std::string &name ) {
    Xrm::opt val = xrm.getResource(name);
    if (!val.first)
        return false;

    return val.second == "True";
}

std::string XOSView::getResourceOrUseDefault( const std::string &name,
  const std::string &defaultVal ){

    Xrm::opt retval = xrm.getResource (name);
    if (retval.first)
        return retval.second;

    return defaultVal;
}

std::string XOSView::getResource( const std::string &name ){
    Xrm::opt retval = xrm.getResource (name);
    if (retval.first)
        return retval.second;
    else {
        logFatal << "Couldn't find '" << name
                 << "' resource in the resource database!\n";
        /*  Some compilers aren't smart enough to know that exit() exits.  */
        return '\0';
    }
}

void XOSView::dumpResources( std::ostream &os ){
    xrm.dump(os);
}
